File: JavaScript/2016/4.js
const lines = require('../getInput')(4, 2016).trim().split('\n');

const count = (arr, needle) => [...arr].filter(char => char === needle).length;
const uniq = arr => arr.reduce((agg, el) => (agg.includes(el) ? agg : [...agg, el]));
const isBeforeInString = s => (a, b) => (count(s, b) - count(s, a) || (a < b ? -1 : 1));
const getTop5 = s => uniq([...s]).sort(isBeforeInString(s)).slice(0, 5);
const re = /(.*)-(\d{3})\[(\w{5})]$/;
const parse = line => ((([, name, id, crc]) => ({ name, id: +id, crc }))(line.match(re)));
const isRealRoom = ({ name, crc }) => crc === getTop5(name.replace(/-/g, '')).join('');
const rotateChar = n => char => String.fromCharCode(97 + (char.charCodeAt() - 97 + n) % 26);
const rotateWord = n => word => [...word].map(rotateChar(n)).join('');
const rotateRoom = ({ id, name }) => ({ name: name.split('-').map(rotateWord(id)).join(' '), id });

const rooms = lines.map(parse).filter(isRealRoom);
const result1 = rooms.reduce((sum, { id }) => sum + id, 0);
const result2 = rooms.map(rotateRoom).filter(({ name, id }) => name.includes('north'));
console.log(result1, result2);

File: Python/2016/4.py
import re

with open('2016/4.txt') as f:
    input = f.read().strip()

reg = r'(.*)-(\d{3})\[(\w{5})]'


def isRoom(room):
    name = room["name"].replace('-', '')
    expectedCrc = sorted(sorted(set(name)),
                         key=lambda char: name.count(char), reverse=True)[:5]
    return ''.join(expectedCrc) == room["crc"]


def parse(line):
    match = re.match(reg, line)
    return {"name": match.group(1), "id": int(match.group(2)),
            "crc": match.group(3)}


def decrypt(room):
    def rotate(word, n):
        return ''.join([chr(97 + ((ord(c) - 97 + n) % 26)) for c in word])
    words = [rotate(word, room["id"]) for word in room["name"].split('-')]
    room["name"] = ' '.join(words)
    return room


def isNorthPole(room):
    return 'north' in room["name"]


rooms = [r for r in [parse(line) for line in input.split('\n')] if isRoom(r)]
print(sum([room["id"] for room in rooms]))
print([room for room in rooms if (isNorthPole(decrypt(room)))])
